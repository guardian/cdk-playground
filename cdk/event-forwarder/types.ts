import type { EventBridgeEvent } from 'aws-lambda';

export type CloudformationEvent = EventBridgeEvent<
	string,
	{
		'stack-id': string;
		'logical-resource-id': string;
		'physical-resource-id': string;
		'resource-type': string;
		'status-details': {
			status: string;
			'detailed-status': string;
			'status-reason': string;
			[additional: string]: string;
		};
	}
>;

export type AutoscalingEvent = EventBridgeEvent<
	string,
	{
		StatusCode: string;
		AutoScalingGroupName: string;
		ActivityId: string;
		Details: Record<string, unknown>;
		RequestId: string;
		EndTime: string;
		EC2InstanceId: string;
		StartTime: string;
		Cause: string;
	}
>;

export type HydratedEvent = (CloudformationEvent | AutoscalingEvent) & {
	cloudformationStackName: string;
};
