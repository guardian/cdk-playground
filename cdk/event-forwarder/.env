STACK=playground
STAGE=DEV
APP=event-forwarder
