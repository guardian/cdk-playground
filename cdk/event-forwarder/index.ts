import {
	AutoScalingClient,
	DescribeAutoScalingGroupsCommand,
} from '@aws-sdk/client-auto-scaling';
import { fromIni } from '@aws-sdk/credential-providers';
import type { AwsCredentialIdentityProvider } from '@smithy/types';
import { getConfig } from './config';
import type {
	AutoscalingEvent,
	CloudformationEvent,
	HydratedEvent,
} from './types';

interface AwsClientConfig {
	region: string;
	credentials?: AwsCredentialIdentityProvider;
}

export async function main(event: CloudformationEvent | AutoscalingEvent) {
	const { stage } = getConfig();

	try {
		const hydrated: HydratedEvent = {
			...event,
			cloudformationStackName: await getCloudformationStackName(event, stage),
		};
		console.log(JSON.stringify(hydrated));
	} catch (err) {
		if (err instanceof Error) {
			console.error(err.message);
		}
		console.log(JSON.stringify(event));
	}
}

async function getCloudformationStackName(
	event: CloudformationEvent | AutoscalingEvent,
	stage: string,
) {
	switch (event.source) {
		case 'aws.autoscaling': {
			return await getCloudformationStackNameForAsg(
				event as AutoscalingEvent,
				stage,
			);
		}
		case 'aws.cloudformation': {
			return getCloudformationStackNameForStack(event as CloudformationEvent);
		}
		default: {
			throw new Error('Unrecognised event source');
		}
	}
}

async function getCloudformationStackNameForAsg(
	event: AutoscalingEvent,
	stage: string,
): Promise<string> {
	const awsConfig: AwsClientConfig = {
		region: 'eu-west-1',
		...(stage === 'DEV' && {
			credentials: fromIni({ profile: 'developerPlayground' }),
		}),
	};

	const { AutoScalingGroupName } = event.detail;

	const client = new AutoScalingClient(awsConfig);
	const command = new DescribeAutoScalingGroupsCommand({
		AutoScalingGroupNames: [AutoScalingGroupName],
	});
	const { AutoScalingGroups = [] } = await client.send(command);

	if (AutoScalingGroups.length !== 1) {
		throw new Error('Unable to locate unique autoscaling group');
	}

	const [cfnStackName] = AutoScalingGroups.flatMap((_) => _.Tags ?? [])
		.filter((_) => _.Key === 'aws:cloudformation:stack-name')
		.map((_) => _.Value);

	if (!cfnStackName) {
		throw new Error(
			'Unable to locate unique tag: aws:cloudformation:stack-name',
		);
	}

	return cfnStackName;
}

async function getCloudformationStackNameForStack(
	event: CloudformationEvent,
): Promise<string> {
	const stackId = event.detail['stack-id'];
	const [, stackName] = stackId.split('/');

	if (!stackName) {
		throw new Error(`Unable to parse stack-id: ${stackId}`);
	}

	return Promise.resolve(stackName);
}
