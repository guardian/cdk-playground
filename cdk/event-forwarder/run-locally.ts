/*
This is a little helper to run a lambda locally.

It purposefully does only one thing to keep it simple.
 */

import type { CloudformationEvent } from './types';
import { main } from './index';

export const sampleCloudformationEvent: CloudformationEvent = {
	version: '0',
	id: 'unknown',
	'detail-type': 'CloudFormation Resource Status Change',
	source: 'aws.cloudformation',
	account: '000000000000',
	time: '2024-08-24T11:38:34Z',
	region: 'eu-west-1',
	resources: [
		'arn:aws:cloudformation:eu-west-1:000000000000:stack/playground-CODE-basic-asg-rolling-update/000000-000000',
	],
	detail: {
		'stack-id':
			'arn:aws:cloudformation:eu-west-1:000000000000:stack/playground-CODE-basic-asg-rolling-update/000000-000000',
		'logical-resource-id': 'AutoScalingGroupBasicASG123',
		'physical-resource-id':
			'playground-CODE-basic-asg-rolling-update-AutoScalingGroupBasicASG123-ABC',
		'resource-type': 'AWS::AutoScaling::AutoScalingGroup',
		'status-details': {
			status: 'UPDATE_IN_PROGRESS',
			'detailed-status': '',
			'status-reason':
				'New instance(s) added to autoscaling group - Waiting on 5 resource signal(s) with a timeout of PT5M.',
		},
	},
};

void main(sampleCloudformationEvent);
