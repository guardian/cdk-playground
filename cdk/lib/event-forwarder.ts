import { GuStack, type GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Arn } from 'aws-cdk-lib';
import { Rule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { LoggingFormat, Runtime } from 'aws-cdk-lib/aws-lambda';

type EventForwarderProps = Omit<GuStackProps, 'stack'>;

export class EventForwarder extends GuStack {
	constructor(scope: App, id: string, props: EventForwarderProps) {
		const app = 'event-forwarder';

		super(scope, id, {
			...props,
			stack: 'deploy',
			app,
			env: {
				region: 'eu-west-1',
			},
		});

		const { account, region } = this;

		const lambda = new GuLambdaFunction(this, 'EventForwarderLambda', {
			app,
			fileName: `${app}.zip`,
			handler: 'index.main',
			runtime: Runtime.NODEJS_20_X,

			/*
      Override the default provided by GuCDK for improved compatability with https://github.com/guardian/cloudwatch-logs-management when producing log lines with markers.
      See also: https://github.com/guardian/cloudwatch-logs-management/issues/326.
       */
			loggingFormat: LoggingFormat.TEXT,
		});

		// This isn't lest privilege, but couldn't get a `condition` to work
		lambda.addToRolePolicy(
			new PolicyStatement({
				effect: Effect.ALLOW,
				actions: ['autoscaling:DescribeAutoScalingGroups'],
				resources: ['*'],
				// conditions: {
				// 	StringEquals: {
				// 		'aws:ResourceTag/gu:repo': this.repositoryName,
				// 	},
				// },
			}),
		);

		new Rule(this, 'CloudformationEventForwarderRule', {
			targets: [new LambdaFunction(lambda)],
			eventPattern: {
				source: ['aws.cloudformation'],
				detailType: ['CloudFormation Resource Status Change'],
				detail: {
					'stack-id': [
						{
							wildcard: Arn.format({
								partition: 'aws',
								service: 'cloudformation',
								region,
								account,
								resource: 'stack',
								resourceName: 'playground-PROD-cdk-playground*',
							}),
						},
					],
				},
			},
		});

		new Rule(this, 'AutoscalingGroupEventForwarderRule', {
			targets: [new LambdaFunction(lambda)],
			eventPattern: {
				source: ['aws.autoscaling'],
				detail: {
					AutoScalingGroupName: [
						{
							wildcard: 'playground-PROD-cdk-playground-*',
						},
					],
				},
			},
		});
	}
}
