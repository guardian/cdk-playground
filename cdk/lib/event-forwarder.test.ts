import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { EventForwarder } from './event-forwarder';

describe('The EventForwarder stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new EventForwarder(app);
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
