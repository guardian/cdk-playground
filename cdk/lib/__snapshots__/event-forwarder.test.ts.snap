// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The EventForwarder stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AutoscalingGroupEventForwarderRuleAllowEventRuleEventForwarderEventForwarderLambda8BE949E1FC259E1D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "EventForwarderLambda9F6516CF",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "AutoscalingGroupEventForwarderRuleE7A6B6BF",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "AutoscalingGroupEventForwarderRuleE7A6B6BF": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "AutoScalingGroupName": [
              {
                "wildcard": "playground-PROD-cdk-playground-*",
              },
            ],
          },
          "source": [
            "aws.autoscaling",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "EventForwarderLambda9F6516CF",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CloudformationEventForwarderRuleAllowEventRuleEventForwarderEventForwarderLambda8BE949E112AE976F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "EventForwarderLambda9F6516CF",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "CloudformationEventForwarderRuleD95AAF62",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CloudformationEventForwarderRuleD95AAF62": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "stack-id": [
              {
                "wildcard": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:cloudformation:eu-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":stack/playground-PROD-cdk-playground*",
                    ],
                  ],
                },
              },
            ],
          },
          "detail-type": [
            "CloudFormation Resource Status Change",
          ],
          "source": [
            "aws.cloudformation",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "EventForwarderLambda9F6516CF",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "EventForwarderLambda9F6516CF": {
      "DependsOn": [
        "EventForwarderLambdaServiceRoleDefaultPolicy5D01F634",
        "EventForwarderLambdaServiceRole25D27763",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "playground/PROD/event-forwarder/event-forwarder.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "event-forwarder",
            "STACK": "playground",
            "STAGE": "PROD",
          },
        },
        "Handler": "index.main",
        "LoggingConfig": {
          "LogFormat": "Text",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "EventForwarderLambdaServiceRole25D27763",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "event-forwarder",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-playground",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "EventForwarderLambdaServiceRole25D27763": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "event-forwarder",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-playground",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EventForwarderLambdaServiceRoleDefaultPolicy5D01F634": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/playground/PROD/event-forwarder/event-forwarder.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/playground/event-forwarder",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/playground/event-forwarder/*",
                  ],
                ],
              },
            },
            {
              "Action": "autoscaling:DescribeAutoScalingGroups",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventForwarderLambdaServiceRoleDefaultPolicy5D01F634",
        "Roles": [
          {
            "Ref": "EventForwarderLambdaServiceRole25D27763",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
